src/Juvix/Backends/LLVM/Net/EAC.hs

```diff
@@ -1,33 +1,193 @@
+-- |
+-- - =EAC= serves as the place where the interaction net rules for the
+--   EAC layer gets run
+-- - The form given to =EAC= is not the base EAC AST, but instead a
+--   pre processed =EAC= graph that the initial graph will be made on
 module Juvix.Backends.LLVM.Net.EAC where
 
 -- TODO ∷ abstract all all imports to LLVM
 
 import qualified Juvix.Backends.LLVM.Codegen as Codegen
 import qualified Juvix.Backends.LLVM.Net.EAC.Types as Types
-import Juvix.Library
+import Juvix.Library hiding (reduce)
+import qualified Juvix.Library.HashMap as Map
+import qualified LLVM.AST as AST
 import qualified LLVM.AST.Constant as C
+import qualified LLVM.AST.IntegerPredicate as IntPred
+import qualified LLVM.AST.Name as Name
 import qualified LLVM.AST.Operand as Operand
 import qualified LLVM.AST.Type as Type
 
 --------------------------------------------------------------------------------
 -- Interaction Net runner
 --------------------------------------------------------------------------------
 
+-- TODO ∷ consider the return type
+-- TODO ∷ remove boileprlate
+reduce ∷
+  ( HasThrow "err" Codegen.Errors m,
+    HasState "blockCount" Int m,
+    HasState "blocks" (Map.HashMap Name.Name Codegen.BlockState) m,
+    HasState "count" Word m,
+    HasState "currentBlock" Name.Name m,
+    HasState "moduleDefinitions" [AST.Definition] m,
+    HasState "names" Codegen.Names m,
+    HasState "symtab" (Map.HashMap Symbol Operand.Operand) m
+  ) ⇒
+  m Operand.Operand
+reduce = Codegen.defineFunction Type.void "reduce" args $
+  do
+    -- recursive function, properly register
+    reduce ← Codegen.externf "reduce"
+    isBothPrimary ← Codegen.externf "is_both_primary"
+    anihilateRewireAux ← Codegen.externf "anihilate_rewire_aux"
+    -- switch creations
+    eacList ← Codegen.externf "eac_list"
+    appCase ← Codegen.addBlock "switch.app"
+    lamCase ← Codegen.addBlock "switch.lam"
+    eraCase ← Codegen.addBlock "switch.era"
+    dupCase ← Codegen.addBlock "switch.dup"
+    defCase ← Codegen.addBlock "switch.default"
+    extCase ← Codegen.addBlock "switch.exit"
+    car ← Types.loadCar eacList
+    cdr ← Types.loadCdr eacList
+    tag ← tagOf car
+    _term ←
+      Codegen.switch
```

Comment:
I wonder if we can do a clever thing where we add the tags with an offset (shift) and have only one case switch - not sure if it would really be faster though.

Reply:
that could be faster, though I'm usnure how LLVM optimizes that

Reply:
yeah let's wait on that sort of thing for now